pallets/parachain-staking/src/lib.rs

```diff
@@ -1312,34 +1903,106 @@ pub mod pallet {
+				return_stake(bond);
+			}
+			// return stake to collator
+			T::Currency::unreserve(&state.id, state.bond);
+			<CandidateState<T>>::remove(&collator);
+			let new_total_staked = <Total<T>>::get().saturating_sub(state.total_backing);
+			<Total<T>>::put(new_total_staked);
+			Self::deposit_event(Event::CollatorLeft(
+				collator,
+				state.total_backing,
+				new_total_staked,
+			));
+			Ok(().into())
+		}
+		#[pallet::weight(0)]
+		/// Cancel open request to leave candidates
+		/// - only callable by collator account
+		/// - result upon successful call is the candidate is active in the candidate pool
+		pub fn cancel_leave_candidates(
+			origin: OriginFor<T>,
+			candidate_count: u32,
+		) -> DispatchResultWithPostInfo {
+			let collator = ensure_signed(origin)?;
+			let mut state = <CandidateState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			ensure!(state.is_leaving(), Error::<T>::CandidateNotLeaving);
+			state.go_online();
+			let mut candidates = <CandidatePool<T>>::get();
+			ensure!(
+				candidates.0.len() as u32 <= candidate_count,
+				Error::<T>::TooLowCandidateCountWeightHintCancelLeaveCandidates
+			);
+			ensure!(
+				candidates.insert(Bond {
+					owner: collator.clone(),
+					amount: state.total_counted
+				}),
+				Error::<T>::AlreadyActive
+			);
+			<CandidatePool<T>>::put(candidates);
+			<CandidateState<T>>::insert(&collator, state);
+			Self::deposit_event(Event::CancelledCandidateExit(collator));
+			Ok(().into())
+		}
 		#[pallet::weight(<T as Config>::WeightInfo::go_offline())]
+		/// Temporarily leave the set of collator candidates without unbonding
 		pub fn go_offline(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let collator = ensure_signed(origin)?;
-			let mut state = <CollatorState2<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			let mut state = <CandidateState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
 			ensure!(state.is_active(), Error::<T>::AlreadyOffline);
 			state.go_offline();
 			let mut candidates = <CandidatePool<T>>::get();
 			if candidates.remove(&Bond::from_owner(collator.clone())) {
 				<CandidatePool<T>>::put(candidates);
 			}
-			<CollatorState2<T>>::insert(&collator, state);
+			<CandidateState<T>>::insert(&collator, state);
 			Self::deposit_event(Event::CollatorWentOffline(
 				<Round<T>>::get().current,
 				collator,
 			));
 			Ok(().into())
 		}
-		/// Rejoin the set of collator candidates if previously had called `go_offline`
 		#[pallet::weight(<T as Config>::WeightInfo::go_online())]
+		/// Rejoin the set of collator candidates if previously had called `go_offline`
 		pub fn go_online(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let collator = ensure_signed(origin)?;
-			let mut state = <CollatorState2<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			let mut state = <CandidateState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
 			ensure!(!state.is_active(), Error::<T>::AlreadyActive);
```

Comment:
This feels weird, cannot we have a collator state called Offline, and just check the collator is in that state?

Reply:
You're right, we could check that it is `state.status == CollatorStatus::Idle`, but because there are three variants, checking that it isn't the other two variants is equivalent.
```rust
#[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
/// The activity status of the collator
pub enum CollatorStatus {
	/// Committed to be online and producing valid blocks (not equivocating)
	Active,
	/// Temporarily inactive and excused for inactivity
	Idle,
	/// Bonded until the inner round
	Leaving(RoundIndex),
}
```
Your suggestion is less code but also a less descriptive error message. It would be `NotOffline` instead of `AlreadyActive` if online and `CannotActBecauseLeaving` if leaving (which is the next line `ensure` check.